File: app/src/main/java/com/evernote/android/job/sample/reminder/ReminderJob.java
package com.evernote.android.job.sample.reminder;

import android.support.annotation.NonNull;

import com.evernote.android.job.Job;
import com.evernote.android.job.JobRequest;
import com.evernote.android.job.util.support.PersistableBundleCompat;

/**
 * @author rwondratschek
 */
public class ReminderJob extends Job {

    public static final String TAG = "ReminderJob";

    private static final String EXTRA_ID = "EXTRA_ID";

    /*package*/ static int schedule(@NonNull Reminder reminder) {
        PersistableBundleCompat extras = new PersistableBundleCompat();
        extras.putInt(EXTRA_ID, reminder.getId());

        long time = Math.max(1L, reminder.getTimestamp() - System.currentTimeMillis());

        return new JobRequest.Builder(TAG)
                .setExact(time)
                .setExtras(extras)
                .setUpdateCurrent(false)
                .build()
                .schedule();
    }

    @NonNull
    @Override
    protected Result onRunJob(Params params) {
        int id = params.getExtras().getInt(EXTRA_ID, -1);
        if (id < 0) {
            return Result.FAILURE;
        }

        Reminder reminder = ReminderEngine.instance().getReminderById(id);
        if (reminder == null) {
            return Result.FAILURE;
        }

        int index = ReminderEngine.instance().getReminders().indexOf(reminder);

        ReminderEngine.instance().showReminder(reminder);
        ReminderEngine.instance().removeReminder(index, false);

        return Result.SUCCESS;
    }
}

File: app/src/main/java/com/evernote/android/job/sample/reminder/ReminderJob.kt
package com.evernote.android.job.sample.reminder

import com.evernote.android.job.Job
import com.evernote.android.job.JobRequest
import com.evernote.android.job.util.support.PersistableBundleCompat

/**
 * @author rwondratschek
 */
class ReminderJob : Job() {

    override fun onRunJob(params: Job.Params): Job.Result {
        val id = params.extras.getInt(EXTRA_ID, -1)
        if (id < 0) {
            return Job.Result.FAILURE
        }

        val reminder = ReminderEngine.getReminderById(id) ?: return Job.Result.FAILURE
        val index = ReminderEngine.getReminders().indexOf(reminder)

        ReminderEngine.showReminder(reminder)
        ReminderEngine.removeReminder(index, false)

        return Job.Result.SUCCESS
    }

    companion object {

        const val TAG = "ReminderJob"
        private const val EXTRA_ID = "EXTRA_ID"

        fun schedule(reminder: Reminder): Int {
            val time = Math.max(1L, reminder.timestamp - System.currentTimeMillis())

            return JobRequest.Builder(TAG)
                .setExact(time)
                .setExtras(PersistableBundleCompat().apply {
                    putInt(EXTRA_ID, reminder.id)
                })
                .setUpdateCurrent(false)
                .build()
                .schedule()
        }
    }
}
